### PR TITLE
Tests for StrawberryType refactor

### PR DESCRIPTION
**Pull request by BryceBeagle**
## Description

This PR is to be merged into #906. This PR is separate in order to better facilitate a review on the testing changes.

TODO: 
- [x] Fix issue with LazyType test 
- [x] Implement the remaining LazyType tests

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tests/types/test_strings.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    assert isinstance(field1.type, StrawberryOptional)
+    assert field1.type.of_type is str
+
+    assert field2.graphql_name == "age"
+    assert isinstance(field2.type, StrawberryOptional)
+    assert field2.type.of_type is int
+
+
+# TODO: Move to object tests to test namespace logic
+def test_basic_list():
```
Comment (BryceBeagle) [inline tests/types/test_strings.py:159]: existing test

File: tests/types/test_strings.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    [field] = definition.fields
+
+    assert field.graphql_name == "names"
+    assert isinstance(field.type, StrawberryList)
+    assert field.type.of_type is str
+
+
+# TODO: Move to object tests to test namespace logic
+def test_list_of_types():
```
Comment (BryceBeagle) [inline tests/types/test_strings.py:175]: existing test

File: tests/types/test_strings.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    assert field1.graphql_name == "name"
+    assert field1.type is str
+
+    assert field2.graphql_name == "age"
+    assert field2.type is int
+
+
+# TODO: Move to object tests to test namespace logic
+def test_optional():
```
Comment (BryceBeagle) [inline tests/types/test_strings.py:138]: existing test

File: tests/types/test_strings.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type is int
+
+    assert resolved == StrawberryOptional(of_type=int)
+    assert resolved == Optional[int]
+
+
+# TODO: Move to object tests to test namespace logic
+def test_basic_types():
```
Comment (BryceBeagle) [inline tests/types/test_strings.py:119]: existing test

File: tests/types/resolving/test_forward_references.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        backward: bool
+
+    resolved = annotation.resolve()
+
+    assert resolved is ForwardClass
+
+    del ForwardClass
+
+
+@pytest.mark.xfail(reason="Combining locals() and globals() strangely makes this fail")
```
Comment (patrick91) [inline tests/types/resolving/test_forward_references.py:23]: is this for another PR? :)

File: tests/types/resolving/test_optionals.py [hunk 6]
```diff
@@ -1,9 +1,65 @@
...
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type == Union[CoolType, UncoolType]
+
+    assert resolved == StrawberryOptional(of_type=Union[CoolType, UncoolType])
+    assert resolved == Optional[Union[CoolType, UncoolType]]
+
+
+# TODO: move to a field test file
```
Comment (patrick91) [inline tests/types/resolving/test_optionals.py:62]: 👀

File: tests/types/resolving/test_unions.py [hunk 7]
```diff
@@ -145,7 +110,7 @@ class B:
 
 def test_error_with_empty_type_list():
     with pytest.raises(TypeError, match="No types passed to `union`"):
-        strawberry.union("Result", [])
+        strawberry.union("Result", ())
```
Comment (patrick91) [inline tests/types/resolving/test_unions.py:113]: nice 😊

File: tests/types/test_object_types.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        name: str
+
+    @strawberry.type
+    class UK:
+        name: str
+
+    EU = strawberry.union("EU", types=(Europe, UK))
+
+    @strawberry.type
+    class WishfulThinking:
```
Comment (patrick91) [inline tests/types/test_object_types.py:122]: 🤣🇪🇺

File: tests/types/test_resolver_types.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        name: str
+
+    resolver = StrawberryResolver(get_sportsball_official)
+    assert resolver.type is FutureUmpire
+
+    del FutureUmpire
+
+
+def test_lazy_type():
+    # TODO: Implement this test
```
Comment (patrick91) [inline tests/types/test_resolver_types.py:40]: 👀

File: tests/types/test_resolver_types.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    @strawberry.type
+    class Diagram:
+        bar: float
+
+    def get_overlap() -> Union[Venn, Diagram]:
+        ...
+
+    resolver = StrawberryResolver(get_overlap)
+    assert resolver.type == Union[Venn, Diagram]
```
Comment (patrick91) [inline tests/types/test_resolver_types.py:104]: do we test calling resolvers?

**Review by patrick91 — APPROVED** [review]

File: tests/types/test_resolver_types.py (hunk 10 above)
Comment (BryceBeagle) [reply tests/types/test_resolver_types.py:104]: I have added no tests calling resolvers. Is there something specific to this change that would warrant additions?

File: tests/types/resolving/test_forward_references.py (hunk 5 above)
Comment (BryceBeagle) [reply tests/types/resolving/test_forward_references.py:23]: Yes, I don't understand why it doesn't work, but it shouldn't affect normal end-users. If you want to take a look, go ahead :slightly_smiling_face:

File: tests/types/test_resolver_types.py (hunk 10 above)
Comment (patrick91) [reply tests/types/test_resolver_types.py:104]: no, just wanted if you wanted to test calling a strawberry resolver directly :)

Comment (BryceBeagle) [reply tests/types/test_resolver_types.py:99]: you mean
```python
resolver = StrawberryResolver(...)
resolver()
```
?

File: tests/types/resolving/test_optionals.py (hunk 6 above)
Comment (BryceBeagle) [reply tests/types/resolving/test_optionals.py:62]: Yeah I didn't want to deal with a complete refactor/reorganize in this PR. Another time

File: tests/types/test_resolver_types.py (hunk 10 above)
Comment (patrick91) [reply tests/types/test_resolver_types.py:99]: yup 😊

Comment (BryceBeagle) [reply tests/types/test_resolver_types.py:99]: I'll open an issue for that, but the logic is very simple.
https://github.com/strawberry-graphql/strawberry/blob/5b601be026a329ed5e04f48b2291b311efa79520/strawberry/types/fields/resolver.py#L21-L22

Comment (BryceBeagle) [reply tests/types/test_resolver_types.py:99]: https://github.com/strawberry-graphql/strawberry/issues/1067

File: tests/types/resolving/test_unions.py [hunk 11]
```diff
@@ -44,49 +40,18 @@ class User:
...
+    assert isinstance(resolved, StrawberryUnion)
+    assert resolved.types == (User, Error)
 
-    strawberry_union = definition.fields[0].type.of_type
-    assert isinstance(strawberry_union, StrawberryUnion)
-    assert strawberry_union.name == "UserError"
-    assert strawberry_union.types == (User, Error)
-
-
-def test_unions_inside_list():
```
Comment (jkimbo) [inline tests/types/resolving/test_unions.py:65]: Why was this test removed?

Comment (BryceBeagle) [reply tests/types/resolving/test_unions.py:65]: Testing lists of unions was moved to the list tests:
https://github.com/strawberry-graphql/strawberry/pull/1063/files/5f33966dc1aa4a2d0b41c1724fd006b9081cc347#diff-7cc80ffd182a1d8dc3ef0e695dd9d4d137b1d45038ce43bc97a9d4627e0f1c68R44-R60

Comment (jkimbo) [reply tests/types/resolving/test_unions.py:65]: 👍

**Review by jkimbo — APPROVED** [review]

**Pull request merged**